Actions on Dashboards 1: Import UI from `data-apps-main`

iethree opened the pull request:
resolves #27738

## Description
Basically imports the necessary pieces from `data-apps-main` to allow users to add action buttons to dashboards and use them. Does not attempt to implement the designs in the [product doc](https://www.notion.so/metabase/Add-buttons-that-trigger-actions-to-dashboards-495ea6b6eabf47fe86f7bb9adcbc57b1)

### Updated
- DashboardHeader: adds an add action menu item for admins
- DashcardActionButton: passes through dashcard property to allow previewing of actions buttons
- Paramaters mapping functions to allow action parameters to be mapped
- Types: mostly `ActionDashboardCard` 

### Added:
- `ActionViz` - the top-level component that handles rendering buttons and forms on dashboards
  - The `ActionOptions` subcomponent which mostly handles parameter mapping will likely be totally replaced by the new action settings modal
- `ActionParamtersInputForm` - the wrapper around our ActionForm component that handles mapping parameters and execution
- `AddActionSidebar` - Allows adding buttons to dashboards
- `ActionPicker` - Allows picking actions from the sidebar - this will be totally replaced by a modal eventually.

## Functionality

You can add actions to dashboards and map dashboard filters to parameters.

![addAndMap](https://user-images.githubusercontent.com/30528226/213298123-61b2b12a-1c73-4991-aae6-b0e7b170554d.gif)

You can execute actions, and linked parameters will get pulled in.

![execute](https://user-images.githubusercontent.com/30528226/213298194-9816b3d2-a318-443c-a381-6d25d08d6edd.gif)

You can add link buttons to dashboards

![linkbuttons](https://user-images.githubusercontent.com/30528226/213298333-298e65ef-b0e2-4831-865e-fbf3cb484a10.gif)

You can edit button appearance and text

![changeSettings](https://user-images.githubusercontent.com/30528226/213298765-43fd5c82-ceaf-4522-86f3-655f40d68835.gif)

You can update an action query

![update actions](https://user-images.githubusercontent.com/30528226/213298813-67d43000-8046-4827-a7a2-fc27a44cbaa0.gif)

iethree committed:
update dashboard + card + parameters to support actions

iethree committed:
update dashboard + card + parameters to support actions

iethree committed:
Add ActionViz Components

iethree committed:
add ActionParameterInputForm

iethree committed:
add ActionParameterInputForm

iethree committed:
Add AddActionSidebar and ActionPicker

kulyk commented on frontend/src/metabase/actions/components/ActionViz/ActionOptions.tsx:
```diff
@@ -0,0 +1,239 @@
+
+export const ActionParameterMappingForm = ({
+  dashcard,
+  page,
```
Can we use `dashboard` here?

kulyk commented on frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:
```diff
@@ -0,0 +1,122 @@
+  dashcard,
+  settings,
+  isSettings,
+  page,
```
Can we use `dashboard` here?

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:
```diff
@@ -0,0 +1,145 @@
+  missingParameters,
+  dashcardParamValues,
+  action,
+  page,
```
Can we use `dashboard` here?

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputModal.tsx:
```diff
@@ -0,0 +1,24 @@
+  children: React.ReactElement;
+}
+
+export default function ActionParametersInputModal({
+  onClose,
+  title,
+  children,
+}: Props) {
+  return (
+    <Modal onClose={onClose}>
+      <ModalContent title={title} onClose={onClose}>
+        {children}
+      </ModalContent>
+    </Modal>
+  );
+}
```
Can we either remove this component or replace `children` with `ActionParametersInputForm`?

kulyk commented on frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:
```diff
@@ -0,0 +1,122 @@
+  VisualizationSettings,
+  Dashboard,
+} from "metabase-types/api";
+import { getFormTitle } from "metabase/actions/containers/ActionCreator/FormCreator";
```
Not sure if it's a big deal, but it feels off that `actions/components` imports something from `actions/containers`

kulyk commented on frontend/src/metabase/actions/components/ActionViz/ActionOptions.tsx:
```diff
@@ -0,0 +1,239 @@
+            <Icon
+              name="bolt"
+              size={14}
+              tooltip={t`Assign action paramter values`}
```
Typo: it should be "parameter" instead of "paramter"

kulyk commented on frontend/src/metabase/actions/components/ActionViz/Action.tsx:
```diff
@@ -0,0 +1,144 @@
+import type { Dispatch } from "metabase-types/store";
+import type { ParameterValueOrArray } from "metabase-types/types/Parameter";
+
+import { generateFieldSettingsFromParameters } from "metabase/actions/containers/ActionCreator/FormCreator";
```
Not sure if that's a big deal, but it feels a bit off that `actions/components` import something from `actions/containers`

kulyk commented on frontend/src/metabase/actions/components/ActionViz/ActionOptions.tsx:
```diff
@@ -0,0 +1,239 @@
+  ParameterMapperContainer,
+  ParameterFormSection,
+  ParameterFormLabel,
+  ParameterMapperTitleContainter,
```
Typo: should be `Container` instead of `Containter` (`ParameterMapperTitleContainer`)

kulyk commented on frontend/src/metabase/actions/components/ActionViz/ActionOptions.tsx:
```diff
@@ -0,0 +1,239 @@
+import Button from "metabase/core/components/Button";
+
+import TippyPopoverWithTrigger from "metabase/components/PopoverWithTrigger/TippyPopoverWithTrigger";
+import ActionCreator from "metabase/actions/containers/ActionCreator";
```
Not sure if that's a big deal, but it feels off that `actions/components` import stuff from `actions/containers`

kulyk reviewed (changes requested):
Looks good! Requesting changes primarily because of `page` occurrences instead of `dashboard` and a few typos; the rest seem non-blocking

kulyk commented:
@iethree can we add some basic tests for important components like `ActionViz` and `ActionInputParameterForm`?

iethree committed:
move ActionParametersInputModal

iethree committed:
move ActionParametersInputModal

iethree committed:
remove page and app references

iethree committed:
reorganize utils

iethree committed:
update types

iethree committed:
check dashboard eligibility for actions access

iethree committed:
check dashboard eligibility for actions access

iethree committed:
fix typos

iethree committed:
remove reference to list viz

iethree committed:
show all models in action picker

iethree committed:
allow delete actions to submit with no form fields

iethree committed:
remove unused testing functions

iethree committed:
add ActionViz unit tests

iethree committed:
add ActionParametersInputForm unit tests

kulyk commented on frontend/src/metabase/actions/components/ActionForm/ActionForm.tsx:
```diff
@@ -175,15 +175,19 @@ export const ActionForm = ({
     >
       {({ dirty }) => (
-        <Form disabled={!dirty} role="form" data-testid="action-form">
+        <Form
+          disabled={!dirty && !!form.fields.length}
+          role="form"
+          data-testid="action-form"
+        >
```
I need a moment to parse `!!form.fields.length`, should we maybe extract it to something like `const hasFields = form.fields.length > 0`? And should we render the form at all if there are no fields?

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:
```diff
@@ -0,0 +1,185 @@
+
+  const shouldPrefetch = useMemo(() => shouldPrefetchValues(action), [action]);
+
+  const fetchInitialValues = useCallback(
+    () =>
+      ActionsApi.prefetchValues({
+        dashboardId: dashboard?.id,
+        dashcardId: dashcard?.id,
+        parameters: JSON.stringify(dashcardParamValues),
+      }).then(setPrefetchValues),
+    [dashboard?.id, dashcard?.id, dashcardParamValues],
+  );
```
Should we use `async/await` here? AFAICT we rarely use promise methods

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:
```diff
@@ -0,0 +1,185 @@
+  );
+
+  useEffect(() => {
+    // we need at least 1 parameter value (a PK) to fetch initial values
+    const canPrefetch =
+      Object.keys(dashcardParamValues).length > 0 && dashboard && dashcard;
```
Can we maybe give `canPrefetch` a different to avoid a need in a comment?

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:
```diff
@@ -0,0 +1,185 @@
+  confirmMessage?: string;
+}
+
+export function ActionParametersInputModal({
+  onClose,
+  title,
+  showConfirmMessage,
+  confirmMessage,
+  ...formProps
+}: ModalProps & Props) {
+  return (
+    <Modal onClose={onClose}>
+      <ModalContent title={title} onClose={onClose}>
+        <>
+          {showConfirmMessage && <ConfirmMessage message={confirmMessage} />}
+          <ActionParametersInputForm {...formProps} />
+        </>
+      </ModalContent>
+    </Modal>
+  );
+}
```
Should we keep it in another file maybe? If I needed to patch it, I'd be looking for a file called `ActionParametersInputModal`

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.unit.spec.tsx:
```diff
@@ -0,0 +1,228 @@
+  ],
+  dashcardParamValues: {},
+  action: createMockQueryAction(),
+  dashboard: { id: 123 },
+  dashcard: { id: 456 },
```
Can we use `createMock` helpers here to make it closer to a real environment?

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.unit.spec.tsx:
```diff
@@ -0,0 +1,228 @@
+  });
+
+  it("should call onCancel when clicking the cancel button", async () => {
+    const cancelSpy = jest.fn();
+    await setup({ onCancel: cancelSpy });
+    await userEvent.click(screen.getByText("Cancel"));
+    expect(cancelSpy).toHaveBeenCalled();
```
Instead of creating mocks per test, what do you think about mocking them inside the `setup` and then return then? Example:

```jsx
function setup({ onCancel = jest.fn() }) {
  render(...)
  return { onCancel }
}

it("should call onCancel", () => {
  const { onCancel } = setup();
  userEvent.click(...)
  expect(onCancel).toHaveBeenCalled();
})
```

iethree replied:
I don't see a problem with that, I was just thinking that I didn't need to create and return a mock for every test when I'm only using it once.

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.unit.spec.tsx:
```diff
@@ -0,0 +1,228 @@
+      onSubmit: submitSpy,
+    });
+
+    await userEvent.type(screen.getByLabelText("Parameter 1"), "uno");
```
Are `userEvent` functions async? I don't think we await them anywhere else

iethree replied:
apparently, only with certain options. https://testing-library.com/docs/ecosystem-user-event/#typeelement-text-options

kulyk commented on frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.unit.spec.tsx:
```diff
@@ -0,0 +1,228 @@
+    });
+}
+
+describe("Actions > ActionParametersInputForm", () => {
```
This test suite throws a `You seem to have overlapping act() calls, this is not supported. Be sure to await previous act() calls before making a new one` error. Could you please investigate? It might be related to awaited `userEvent` calls or maybe async functions inside `waitFor`

iethree replied:
looks like it was a consequence of using `await findByX()` within `await watiFor()`

kulyk commented on frontend/src/metabase/dashboard/actions/data-fetching.js:
```diff
@@ -108,10 +108,7 @@ const loadingComplete = createThunkAction(
-        ? t`Your page is ready`
-        : t`Your dashboard is ready`;
+      const message = t`Your dashboard is ready`;
       dispatch(setDocumentTitle(message));
```
nit: think there's no point in the `message` variable anymore and we should be able to just pass it to `setDocumentTitle`

kulyk commented on frontend/src/metabase/dashboard/components/AddActionSidebar/AddActionSidebar.tsx:
```diff
@@ -0,0 +1,120 @@
+
+export const AddActionSidebar = _.compose(
+  Search.loadList({
+    query: (_state: State, props: any) => ({
+      models: ["dataset"],
+    }),
```
Think we can remove an unused `props` argument here. Minus one `any` :D

kulyk commented on frontend/src/metabase/nav/containers/MainNavbar/types.ts:
```diff
@@ -17,13 +17,7 @@ export interface MainNavbarDispatchProps {
 export type MainNavbarProps = MainNavbarOwnProps & MainNavbarDispatchProps;
 
 export interface SelectedItem {
-  type:
-    | "card"
-    | "collection"
-    | "dashboard"
-    | "data-app"
-    | "data-app-page"
-    | "non-entity";
```
❤️

kulyk commented on frontend/src/metabase/nav/utils/model-names.ts:
```diff
@@ -1,7 +1,6 @@
 import { t } from "ttag";
 
 const TRANSLATED_NAME_BY_MODEL_TYPE: Record<string, string> = {
-  app: t`App`,
```
❤️

kulyk commented on frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx:
```diff
@@ -0,0 +1,242 @@
+      },
+    ],
+  },
+  dashboard: { id: 123 },
```
Can we use `createMockDashboard` utility to make it closer to a real environment?

iethree commented on frontend/src/metabase/actions/components/ActionViz/ActionOptions.tsx:
```diff
@@ -0,0 +1,239 @@
+  };
+};
+
+type ParamaterMappingFn = (
```
this is the hardest word to spell in the whole world 🤦‍♀️

kulyk committed:
Fix native query editor mock in `ActionCreator` tests (#27794)

* Update `NativeQueryEditor` mock

* Mock `NativeQueryEditor` before registering viz

iethree committed:
more cleanup

iethree committed:
add more tests

kulyk commented on frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx:
```diff
@@ -0,0 +1,286 @@
+        },
+      };
+
+      const scope = nock(location.origin)
+        .post("/api/dashboard/123/dashcard/456/execute", expectedBody)
+        .reply(200, { "rows-updated": [1] });
```
I think it's [preferred](https://www.notion.so/metabase/46-Document-the-pattern-with-setup-function-in-jest-tests-ce3c9f374321427dabe0f26e719aecee) to mock server endpoints inside the `setup`

kulyk commented on frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx:
```diff
@@ -0,0 +1,286 @@
+        linkType: "url",
+        linkTemplate: "https://metabase.com",
+      };
+      const { container } = await setup({
```
Doesn't look like we need this `container` here

kulyk commented on frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx:
```diff
@@ -0,0 +1,286 @@
+        linkTemplate: "https://metabase.com",
+      };
+      const { container } = await setup({
+        dashcard: { action: undefined },
```
This dashcard seems invalid? It's basically an empty object here?

kulyk commented on frontend/src/metabase/actions/components/ActionViz/Action.unit.spec.tsx:
```diff
@@ -0,0 +1,286 @@
+        linkType: "url",
+        linkTemplate: "https://metabase.com",
+      };
+      const { container } = await setup({
```
Doesn't look like we need this `container` here

kulyk reviewed (approved):
Couple of comments around testing, but overall LGTM 🎉 Awesome work! Thanks for addressing all of the comments on that

iethree committed:
move api mock to setup function

iethree commented:
> Couple of comments around testing, but overall LGTM 🎉 Awesome work! Thanks for addressing all of the comments on that

thank *you*. I really appreciate the thorough review!
